### PR TITLE
Psd 65/filter invulnerability status

### PR DESCRIPTION
**Pull request by fakebrandonchan**
Added a filter that filters on status
-status is now a part of the incoming payload
-filtering around status is now included
-added a test to check that this works

### CODE REVIEW

File: pkg/filter/filter.go [hunk 1]
```diff
@@ -69,13 +75,22 @@ func (f VulnerabilityFilter) FilterVulnerabilities(ctx context.Context, asset do
...
 			filteredVulnerabilities = append(filteredVulnerabilities, vuln)
 			logger.Info(logs.VulnerabilityFiltered{
 				Action:  logs.VulnRetained,
 				Method:  logs.CvssV2Score,
 				VulnID:  vuln.ID,
 				AssetID: asset.ID,
+				Status:  vuln.Status,
 			})
 			stater.Count("event.nexposevulnerability.filter.accepted", 1)
 		case f.VulnIDRegexp.MatchString(vuln.ID):
```
Comment (fakebrandonchan) [inline pkg/filter/filter.go:105]: why are we checking this? what is the purpose of this case, to check the ID

File: pkg/filter/filter.go [hunk 2]
```diff
@@ -69,13 +75,22 @@ func (f VulnerabilityFilter) FilterVulnerabilities(ctx context.Context, asset do
 	filteredVulnerabilities := make([]domain.Vulnerability, 0)
 	for _, vuln := range vulnerabilities {
 		switch {
+		case vuln.Status == invulnerable || vuln.Status == noResults:
```
Comment (fakebrandonchan) [inline pkg/filter/filter.go]: I'm uncertain of where to put this case block in the switch statement. I also feel like there's a lot of duplicate code, at least the log actions to me looks like duplicate work albeit a couple things

File: pkg/filter/filter_test.go [hunk 3]
```diff
@@ -95,6 +95,24 @@ func TestFilterVulnerabilities(t *testing.T) {
...
+			"filters out vulnerabilities that have bad status",
+			7.0,
+			"test-vuln-.*",
+			[]domain.Vulnerability{
+				domain.Vulnerability{
+					ID:          "bad-vuln-1",
+					CvssV2Score: 8.0,
+					Status:      "invulnerable",
+				},
+				// domain.Vulnerability{
```
Comment (bkrebsbach) [inline pkg/filter/filter_test.go]: Task: This should be uncommented or removed.

File: pkg/filter/filter.go [hunk 4]
```diff
@@ -85,15 +100,17 @@ func (f VulnerabilityFilter) FilterVulnerabilities(ctx context.Context, asset do
...
 			stater.Count("event.nexposevulnerability.filter.accepted", 1)
 		default:
 			logger.Info(logs.VulnerabilityFiltered{
 				Action:  logs.VulnDiscarded,
 				VulnID:  vuln.ID,
 				AssetID: asset.ID,
+				Status:  vuln.Status,
 			})
-			stater.Count("event.nexposevulnerability.filter.discarded", 1)
+			stater.Count("event.nexposevulnerability.filter.discarded", 1, fmt.Sprintf("reason:%s", "default"))
```
Comment (bkrebsbach) [inline pkg/filter/filter.go]: Suggestion: I'd change `"default"` to `"unknown"` and make it a constant.

File: pkg/filter/filter.go (hunk 1 above)
Comment (bkrebsbach) [reply pkg/filter/filter.go:105]: This adds vulnerabilities which match the acceptable vulnerability ID regex to the list of vulns that are passed to the next service in the pipeline.

File: pkg/filter/filter.go [hunk 5]
```diff
@@ -69,13 +75,22 @@ func (f VulnerabilityFilter) FilterVulnerabilities(ctx context.Context, asset do
...
 	for _, vuln := range vulnerabilities {
 		switch {
+		case vuln.Status == invulnerable || vuln.Status == noResults:
+			logger.Info(logs.VulnerabilityFiltered{
+				Action:  logs.VulnDiscarded,
+				VulnID:  vuln.ID,
+				AssetID: asset.ID,
+				Status:  vuln.Status,
+			})
+			stater.Count("event.nexposevulnerability.filter.discarded", 1, fmt.Sprintf("reason:%s", "status"))
```
Comment (bkrebsbach) [inline pkg/filter/filter.go]: Suggestion: I'd split this into two cases, and change "status" to "invulnerable" or "no-results", making use of the constants.

File: gateway-incoming.yaml [hunk 6]
```diff
@@ -157,6 +157,10 @@ components:
             independent results.
           items:
             $ref: '#/components/schemas/AssessmentResult'
+        status:
+          type: string
+          example: vulnerable
+          description: variable to filter on
```
Comment (bkrebsbach) [inline gateway-incoming.yaml]: Suggestion: this doesn't really make clear what `status` means in this context. I'd recommend using the description from your nexpose-vuln-hydrator PR for clarity and consistency.

File: pkg/filter/filter_test.go (hunk 3 above)
Comment (fakebrandonchan) [reply pkg/filter/filter_test.go]: fixed

File: pkg/filter/filter.go (hunk 4 above)
Comment (fakebrandonchan) [reply pkg/filter/filter.go]: fixed

File: pkg/filter/filter.go (hunk 5 above)
Comment (fakebrandonchan) [reply pkg/filter/filter.go]: fixed

File: gateway-incoming.yaml (hunk 6 above)
Comment (fakebrandonchan) [reply gateway-incoming.yaml]: fixed

**Review by bkrebsbach — DISMISSED** [review]

**Review by faldridge — DISMISSED** [review]
Great job! This looks perfect.

**Review by faldridge — APPROVED** [review]

**Review by mikerott — APPROVED** [review]
This looks complete to me.

**Pull request merged**